src/runtime_src/core/common/api/xrt_error.cpp

```diff
@@ -205,6 +206,27 @@ class error_impl
 public:
   error_impl(const xrt_core::device* device, xrtErrorClass ecl)
   {
+    //Code for new format; Binary array of error structs in sysfs
+    try {
+      std::vector<char> buf = xrt_core::device_query<xrt_core::query::xocl_errors>(device);
+      if (buf.empty())
+        return;
+      const xcl_errors *errors_buf = reinterpret_cast<xcl_errors *>(buf.data());
+      if (errors_buf->num_err <= 0)
+        return;
+
+      for (int i = 0; i < errors_buf->num_err; i++) {
+        if (XRT_ERROR_CLASS(errors_buf->errors[i].err_code) != ecl)
+          continue;
+        m_errcode = XRT_ERROR_NUM(errors_buf->errors[i].err_code);
+        m_timestamp = errors_buf->errors[i].ts;
+      }
+      return;
+    } catch (const xrt_core::query::no_such_key&) {
+      // Ignoring for now. Check below for edge if not available
+      // query table of zocl doesn't have xocl_errors key
```

Comment:
This is a parsing routine specific to the query request.  I don't think this code should do reinterpret_cast of the query request return value as it is details that are specific to the xrt_core::query::xocl_errors query request.

Use the `struct xrt_core::query::xocl_errors` in query_requests.h to define your parsing of what the query request returns.

The code in this file should preferably not interpret raw data.  You should call the query request as you do, but then call a parsing routine in `struct xrt_core::query::xocl_errors` that interprets the raw data and validates it also.

There is nothing critical in this code, so I think a parsing routine in the struct could return a std::vector of pairs which would then leave the original code almost intact.

The new code is not checking for multiple errors for specified error class, it is just using the last. Maybe this is fine, but it is a change.

Reply:
Done.
Yes change in behavior as latest entry will be used which will be at the end of the array.
Multiple entries to be allowed